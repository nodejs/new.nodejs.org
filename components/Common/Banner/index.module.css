.banner {
  @apply flex
    w-full
    flex-row
    items-center
    justify-center
    gap-2
    px-8
    py-3
    text-sm;

  &,
  a {
    @apply text-white
      dark:text-white;
  }

  a {
    @apply w-fit
      underline
      decoration-white-opaque;
  }

  svg {
    @apply h-4
      w-4
      text-white-opaque;
  }
}

.default {
  @apply bg-green-600;
}

.error {
  @apply bg-danger-600;
}

.warning {
  @apply bg-warning-600;
}
