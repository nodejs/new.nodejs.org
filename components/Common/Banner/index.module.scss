@import 'styles/_variables';

.banner {
  color: $node-gray;
  margin: 0 auto;
  max-width: 90vw;
  padding-top: var(--space-08);
  position: relative;

  &,
  p {
    font-size: var(--font-size-body1);
    font-weight: var(--font-weight-bold);
  }

  a {
    align-items: center;
    display: flex;
    flex-direction: column;
    text-decoration: none;

    &:hover {
      background-color: transparent !important;
      color: $node-green;
      text-decoration: underline;
    }

    img {
      border-radius: 5px;
      max-width: 100%;
      object-fit: cover;
    }

    &.bannerBtn {
      background: var(--purple5);
      padding: 0 var(--space-16);
      border: 1px solid transparent;
      border-radius: 5.6rem;
      color: var(--color-fill-top-nav);
      font-family: var(--sans);
      font-style: normal;
      font-weight: var(--font-weight-semibold);
      line-height: var(--line-height-subheading);
      margin-right: var(--space-32);
      position: relative;
      text-decoration: none;
      white-space: nowrap;
      font-size: 1rem;

      &:hover {
        cursor: pointer;
        background-color: $node-green !important;
      }
    }
  }

  p {
    align-items: center;
    background-color: var(--color-fill-banner);
    border-radius: 5px;
    display: flex;
    flex-direction: row;
    margin: 0;
    padding: var(--space-12);
    text-align: center;

    a {
      &:hover {
        text-decoration: underline;
        background-color: transparent !important;
      }
    }
  }
}
