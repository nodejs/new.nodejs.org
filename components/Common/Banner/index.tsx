import LocalizedLink from '@/components/LocalizedLink';
import { ArrowUpRightIcon } from '@heroicons/react/24/outline';
import styles from './index.module.scss';
import type { FC } from 'react';

type BannerProps = {
  type: 'default' | 'error' | 'warning';
  text: string;
  url?: string;
};

const Banner: FC<BannerProps> = ({ type, text, url = '' }) => (
  <div className={`${styles.banner} ${styles[type] || styles.default}`}>
    {(url.length > 0 && <LocalizedLink href={url}>{text}</LocalizedLink>) ||
      text}
    {url.length > 0 && <ArrowUpRightIcon />}
  </div>
);

export default Banner;
