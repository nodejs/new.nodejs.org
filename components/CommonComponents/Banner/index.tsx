import React, { useMemo } from 'react';
import { sanitize } from 'isomorphic-dompurify';
import { dateIsBetween } from '../../../util/dateIsBetween';
import { isAbsoluteUrl } from '../../../util/isAbsoluteUrl';
import styles from './index.module.scss';
import { WebsiteBanner } from '../../../types';

export interface BannerProps {
  bannersIndex: WebsiteBanner;
}

const useTextContent = ({ text, link, bannerBtnText }: WebsiteBanner) =>
  useMemo(() => {
    if (text) {
      return (
        <p>
          <a href={link} target="_blank" rel="noopener noreferrer">
            <button type="button">{bannerBtnText || 'Read More'}</button>
          </a>
          {text}
        </p>
      );
    }

    return null;
  }, [text, link, bannerBtnText]);

const useHtmlContent = ({ html, link }: WebsiteBanner) =>
  useMemo(() => {
    if (html) {
      const sanitizedHtml = sanitize(html);

      return (
        <a
          href={link}
          target="_blank"
          rel="noopener noreferrer"
          dangerouslySetInnerHTML={{ __html: sanitizedHtml }}
        />
      );
    }

    return null;
  }, [html, link]);

const Banner = ({ bannersIndex }: BannerProps) => {
  const showBanner = dateIsBetween(
    bannersIndex.startDate,
    bannersIndex.endDate
  );

  const link = !isAbsoluteUrl(bannersIndex.link)
    ? `http://nodejs.org/${bannersIndex.link}`
    : bannersIndex.link;

  const textContent = useTextContent({ ...bannersIndex, link });
  const htmlContent = useHtmlContent({ ...bannersIndex, link });

  if (showBanner) {
    return (
      <div className={styles.banner}>
        {bannersIndex.text ? textContent : htmlContent}
      </div>
    );
  }

  return null;
};

export default Banner;
