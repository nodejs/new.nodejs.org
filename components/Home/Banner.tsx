import { siteConfig } from '@/next.json.mjs';
import { dateIsBetween } from '@/util/dateIsBetween';

const Banner = () => {
  if (siteConfig.websiteBanners && siteConfig.websiteBanners['index']) {
    const indexBanner = siteConfig.websiteBanners['index'];

    const showBanner = dateIsBetween(
      indexBanner.startDate,
      indexBanner.endDate
    );

    if (showBanner && indexBanner.text) {
      return (
        <p className="home-version home-version-banner">
          <a href={indexBanner.link}>{indexBanner.text}</a>
        </p>
      );
    }

    if (showBanner && indexBanner.html) {
      return (
        <a
          href={indexBanner.link}
          dangerouslySetInnerHTML={{ __html: indexBanner.html }}
        />
      );
    }
  }

  return null;
};

export default Banner;
