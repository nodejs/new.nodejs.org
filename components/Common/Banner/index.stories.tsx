import Banner from './index';
import type { Meta as MetaObj, StoryObj } from '@storybook/react';

type Story = StoryObj<typeof Banner>;
type Meta = MetaObj<typeof Banner>;

const addDaysToDate = (numDays: number, date: Date) => {
  const newDate = new Date(date);
  newDate.setDate(date.getDate() + numDays);
  return newDate;
};

// Create mock start and end dates as Banner Component renders
// only if end date is on or after today's date
const startDate = new Date();
const endDate = addDaysToDate(3, startDate);

export const WithText: Story = {
  args: {
    bannersIndex: {
      startDate: startDate.toISOString(),
      endDate: endDate.toISOString(),
      text: 'Banner Text',
      link: 'https://nodejs.org/en/',
    },
  },
};

export const WithHTML: Story = {
  args: {
    bannersIndex: {
      startDate: startDate.toISOString(),
      endDate: endDate.toISOString(),
      html: '<p>Banner HTML</p>',
      link: 'https://nodejs.org/en/',
    },
  },
};

export const WithHTMLImage: Story = {
  args: {
    bannersIndex: {
      startDate: startDate.toISOString(),
      endDate: endDate.toISOString(),
      html: '<img src="/static/images/nodejs-training.png" alt="Banner Image" />',
      link: 'https://nodejs.org/en/',
    },
  },
};

export default { component: Banner } as Meta;
