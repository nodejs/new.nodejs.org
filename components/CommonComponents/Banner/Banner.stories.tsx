import type { Meta, StoryObj } from '@storybook/react';
import Banner from './';

const meta: Meta<typeof Banner> = {
  title: 'Banner',
  component: Banner,
};

export default meta;
type Story = StoryObj<typeof Banner>;

const startDate = new Date();
const endDate = new Date();
endDate.setDate(endDate.getDate() + 3);

export const WithText: Story = {
  args: {
    bannersIndex: {
      startDate: startDate.toISOString(),
      endDate: endDate.toISOString(),
      text: 'Banner Text',
      link: 'https://nodejs.org/en/',
      bannerBtnText: 'Read more',
    },
  },
};

export const WithHTML: Story = {
  args: {
    bannersIndex: {
      startDate: startDate.toISOString(),
      endDate: endDate.toISOString(),
      html: '<p>Banner HTML</p>',
      link: 'https://nodejs.org/en/',
    },
  },
};
