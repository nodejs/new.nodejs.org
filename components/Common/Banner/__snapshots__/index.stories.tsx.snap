// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Common/Banner WithHTML smoke-test 1`] = `
<div class="Banner_banner__GAFb9">
  <a href="https://nodejs.org/en/"
     target="_blank"
     rel="noopener noreferrer"
  >
    <p>
      Banner HTML
    </p>
  </a>
</div>
`;

exports[`Common/Banner WithHTMLImage smoke-test 1`] = `
<div class="Banner_banner__GAFb9">
  <a href="https://nodejs.org/en/"
     target="_blank"
     rel="noopener noreferrer"
  >
    <img src="/static/images/nodejs-training.png"
         alt="Banner Image"
    >
  </a>
</div>
`;

exports[`Common/Banner WithText smoke-test 1`] = `
<div class="Banner_banner__GAFb9">
  <p>
    <a href="https://nodejs.org/en/"
       class="Banner_bannerBtn__FqghI"
       target="_blank"
       rel="noopener noreferrer"
    >
      Read More
    </a>
    Banner Text
  </p>
</div>
`;
